File: src/importance.R
# packages and libraries --------------------------------------------------
require(randomForest)
library(pbapply)
library(data.table)
library(parallel)


HOLDOUT_NUM = 10
SEED = 15
set.seed(SEED)
CORES_NUM = 10 #min(25,int(os.cpu_count()))
PAR = TRUE
RESPONSE_VAR = 'y'

# MODEL <- 'r_rf_pyParams'
MODEL <- 'r_rf_default'
dataset <- 'concrete'

# setwd("C:/Users/tommc/OneDrive/Code/predictiveModelling/R_vs_Python/")
setwd('B:/OneDrive/Code/predictiveModelling/R_vs_Python')
DATA_PATH = paste0('data/data_',dataset,'.csv')
data <- read.csv(DATA_PATH, stringsAsFactors=FALSE)
# Read model
# rf <- readRDS(paste0('results/',dataset,'/models/r_rf_default.rds'))
rf <- randomForest(y~.,data=data)

# import data
DATA_PATH = paste0('data/data_',dataset,'.csv')
data <- read.csv(DATA_PATH)

if (DATA_PATH=='data/data_zeroinflate.csv'){
  # Factor variables
  data[,c('data', 'x49','x50','x51','x52','x53','x54')] <- lapply(data[, c('x48', 'x49','x50','x51','x52','x53','x54')], as.factor)
} else if (DATA_PATH=='data/data_lst.csv') {
  # all numeric
  data <- as.data.frame(sapply( data, as.numeric ))
  data <- data[complete.cases(data),]
}

# importance plot
jpeg(paste0("fig/varimp_R_",dataset,".jpg"), width = 1280, height = 960, units = "px", pointsize=18)
varImpPlot(rf, type=2, main=paste0('R Feature Importance: ',dataset))

dev.off()


# partial dependence
imp = importance(rf)
imp_vars = rownames(imp)[order(imp,decreasing = T)[1:4]]

jpeg(paste0("fig/pdp_R_",dataset,".jpg"), width = 1280, height = 960, units = "px", pointsize=18)
op <- par(mfrow=c(2, 2))
for (i in seq_along(imp_vars)){
  partialPlot(rf, data, imp_vars[i], xlab=imp_vars[i],
                main='',
                ylim=c(30, 70))
}
dev.off()

File: src/importance.py
from sklearn.ensemble.partial_dependence import partial_dependence
from sklearn.ensemble.partial_dependence import plot_partial_dependence
import os
import pandas as pd
import numpy as np
import csv
import matplotlib.pyplot as plt
from scipy import stats
from joblib import Parallel, delayed
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV, ShuffleSplit
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.linear_model import SGDClassifier, LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.neural_network import MLPClassifier
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from xgboost import XGBRegressor
import time
import code

dataset = 'concrete'

# define constants
DATA_PATH = 'data/data_{}.csv'.format(dataset)
TIME_PATH = 'data/time_elapsed_{}.csv'.format(dataset)
SEED = 15
RESPONSE_VAR = 'y'

def main():
    # loop different models
    # import the data
    data = import_data()
    # fit model
    forest_reg, x_train = py_rf_rParams(data)
    # get the variable importance
    indices,vars = plot_importances(forest_reg,data)
    # get the partial dependence
    plot_dependence(data)

def import_data():
    # import the data and separate the test and train
    # import data
    data = pd.read_csv(DATA_PATH)
    data = data.dropna()
    # drop columns
    # data = data.drop(['x6','x7','x9','x10'], axis=1)
    if DATA_PATH == 'data/data_zeroinflate.csv':
        # convert categorical variables
        var_cat = ['x48', 'x49','x50','x51','x52','x53','x54']
        for v in var_cat:
            data[v] = pd.factorize(data[v])[0]
    else:
        data = data.apply(pd.to_numeric)
        data = data.dropna()
    return(data)


def py_rf_rParams(data):
    # Random forest
    # data split
    x_train = data.copy()
    y_train = data[RESPONSE_VAR].copy()
    x_train = x_train.drop(RESPONSE_VAR, axis=1)
    # train
    forest_reg = RandomForestRegressor(random_state=SEED, n_estimators=500, max_features=1/3)
    forest_reg.fit(x_train, y_train)
    return(forest_reg,x_train)


def plot_importances(forest_reg,data):
    importances = forest_reg.feature_importances_
    # num_to_show = np.min(10,len(data.columns))
    indices = np.argsort(importances)[::-1]
    # import code
    # code.interact(local=locals())
    vars = data.columns[indices]
    plt.bar(range(len(indices)), importances[indices])
    plt.title('Python Feature Importance: {}'.format(dataset))
    plt.xticks(range(len(indices)), vars, rotation=90)
    plt.xlabel('')
    plt.xlim([-1,len(indices)])
    plt.tight_layout()
    plt.savefig('fig/varimp_py_{}.png'.format(dataset), format='png', dpi=200, transparent=False)
    # plt.show()
    plt.clf()
    return(indices,vars)

def plot_dependence(data):
    ''' Plot the partial dependence '''
    # train a gbm
    x_train = data.copy()
    y_train = data[RESPONSE_VAR].copy()
    x_train = x_train.drop(RESPONSE_VAR, axis=1)
    # train
    reg = GradientBoostingRegressor(random_state=SEED, n_estimators=500, max_features=1/3)
    reg.fit(x_train, y_train)

    # determine importances
    importances = reg.feature_importances_
    indices = np.argsort(importances)[::-1]
    var_names = x_train.columns[indices]

    # partial dependence
    features = list(indices[0:4])
    names = list(var_names[0:4])
    # import code
    # code.interact(local=locals())
    fig, axs = plot_partial_dependence(reg, x_train, features,
                                       feature_names=x_train.columns,
                                       n_jobs=3, grid_resolution=50,
                                       n_cols=2)
    plt.tight_layout()  # tight_layout causes overlap with suptitle
    plt.savefig('fig/pdp_py_{}.png'.format(dataset), format='png', dpi=200, transparent=False)
    plt.show()

if __name__ == '__main__':
    main()
